### ISSUE TITLE
[1.0] Datagrid pagination emits clrDgRefresh twice

### ISSUE THREAD

**Issue by whuang02** [vmware/clarity#2920]
Hi,

I'm trying to get server driven pagination to work, but everytime I change pages, the refresh is emitted twice causing issues:

page: {from: 1, to: 1, size: 1} // Emitted page state when I click next page
page: {from: 0, to: 0, size: 1} // a second refresh emitted with following page state

```
      <clr-datagrid (clrDgRefresh)="onRefresh($event)" [clrDgLoading]="loading">
            <clr-dg-action-bar>    
                <button class="btn btn-sm" (click)="onAdd()">
                    <clr-icon shape="plus"></clr-icon> Add
                </button>
            </clr-dg-action-bar>
            <clr-dg-column [clrDgField]="'name'">DataItem</clr-dg-column>
            <clr-dg-column></clr-dg-column>
            <clr-dg-placeholder>We couldn't find any data!</clr-dg-placeholder>
            <clr-dg-row *clrDgItems="let dataItem of gridData" [clrDgItem]="dataItem">
                <clr-dg-cell>{{dataItem.name}}</clr-dg-cell>
            </clr-dg-row>
            <clr-dg-footer>
                <clr-dg-pagination #pagination [clrDgTotalItems]="totalItems" [clrDgPageSize]="1">
                    <clr-dg-page-size [clrPageSizeOptions]="[1, 25, 50, 100]">Items per page</clr-dg-page-size>
                    {{pagination.firstItem + 1}} - {{pagination.lastItem + 1}} of {{pagination.totalItems}}
                </clr-dg-pagination>
            </clr-dg-footer>
        </clr-datagrid>
```
  ```
 protected onRefresh(state: ClrDatagridStateInterface): void {
        console.log("refreshed state");
        this.loadData(state);
    }

   private loadData(state: ClrDatagridStateInterface){
        this.dataService.read(state)
            .pipe(
                tap(() =>{
                    this.loading = true;
                }),
                finalize(() =>{
                    this.loading = false;
                })
            )
            .subscribe(
                (response) => {
                    this.gridData = response.content;
                    this.totalItems = response.totalElements;
                },
                (error: HttpErrorResponse) => {
                    console.log("failed to load data");
                }
            );
    }
```

Is this a bug or am I using the pagination incorrectly?

**Comment by youdz** [comment]
You are using `*clrDgItems` with a server-driven datagrid, which is why it breaks. If you only give Clarity the items for the currently displayed page, you need to use `*ngFor` as is explained in [our documentation](https://vmware.github.io/clarity/documentation/v1.0/datagrid/server-driven): 
>One important thing to note is that since you don't have all the data available at once, you cannot use the smart iterator *clrDgItems: it would sort, filter and paginate a subset of the data that has already gone through all that. So all you need to do is to go back to a simple *ngFor, which we support.

**Comment by hamza1100** [comment]
@youdz i'm having the same issue. after changing it from *clrDgItems to *ngFor it called refresh function only once but I also need the sorting and filtering, and for that *ngFor does not work. can you please give me a solution for this problem?

**Comment by hippee-lee** [comment]
Hi @hamza1100 - this issue was closed quite a while ago. Can you open a new issue with a stackblitz that demonstrates the issue you are encountering?

I know that it seems similar but the internal datagrid code has changed quite a bit since this issue is closed so it will be helpful to the investigation if there is an up to date reproduction.
